examples/levy-ssm/script.jl

```diff
@@ -0,0 +1,257 @@
+
+function integral(times::Array{Float64}, path::GammaPath)
+    let jumps = path.jumps, jump_times = path.times
+        return [sum(jumps[jump_times .<= t]) for t in times]
+    end
+end
+
+# Gamma Process
+C = 1.0
+beta = 1.0
+ϵ = 1e-10
+process = GammaProcess(C, beta)
+
+# Normal Mean-Variance representation
+μw = 0.0
+σw = 1.0
+nvm = NormalMeanVariance(μw, σw)
+
+# Levy SSM with Langevin dynamics
+#   dx(t) = A x(t) dt + L dW(t)
+#   y(t)  = H x(t) + ϵ(t)
+θ = -0.5
+A = [
+    0.0 1.0
+    0.0 θ
+]
+L = [0.0; 1.0]
+σe = 1.0
+H = [1.0, 0]
+dyn = LangevinDynamics(A, L, θ, H, σe)
+
+# Simulation parameters
+start, finish = 0, 100
+N = 200
+ts = range(start, finish; length=N)
+seed = 4
+rng = Random.MersenneTwister(seed)
+Np = 10
+Ns = 10
+
+f(dt, θ) = exp(θ * dt)
+function Base.exp(dyn::LangevinDynamics, dt::Real)
+    let θ = dyn.θ
+        f_val = f(dt, θ)
+        return [1.0 (f_val - 1)/θ; 0 f_val]
+    end
+end
+
+function meancov(
+    t::T, dyn::LangevinDynamics, path::GammaPath, nvm::NormalMeanVariance
+) where {T<:Real}
+    μ = zeros(T, 2)
+    Σ = zeros(T, (2, 2))
+    let times = path.times, jumps = path.jumps, μw = nvm.μ, σw = nvm.σ
+        for (v, z) in zip(times, jumps)
+            ft = exp(dyn, (t - v)) * dyn.L
+            μ += ft .* μw .* z
+            Σ += ft * transpose(ft) .* σw^2 .* z
+        end
+        return μ, Σ
+    end
+end
+
+X = zeros(Float64, (N, 2))
+Y = zeros(Float64, N)
+for (i, t) in enumerate(ts)
+    if i > 1
+        s = ts[i - 1]
+        dt = t - s
+        path = simulate(process, dt, s, t, ϵ)
+        μ, Σ = meancov(t, dyn, path, nvm)
+        X[i, :] .= rand(rng, MultivariateNormal(exp(dyn, dt) * X[i - 1, :] + μ, Σ))
+    end
+
+    let H = dyn.H, σe = dyn.σe
+        Y[i] = transpose(H) * X[i, :] + rand(rng, Normal(0, σe))
+    end
+end
+
+# AdvancedPS
+Parameters = @NamedTuple begin
+    dyn::LangevinDynamics
+    process::GammaProcess
+    nvm::NormalMeanVariance
+    times::Vector{Float64}
+end
+
+struct MixedState{T}
+    x::Vector{T}
+    path::GammaPath{T}
+end
+
+mutable struct LevyLangevin <: AdvancedPS.AbstractStateSpaceModel
+    X::Vector{MixedState{Float64}}
+    θ::Parameters
+    LevyLangevin(θ::Parameters) = new(Vector{MixedState{Float64}}(), θ)
+end
+
+struct InitialDistribution end
+function Base.rand(rng::Random.AbstractRNG, ::InitialDistribution)
+    return MixedState(rand(MultivariateNormal([0, 0], I)), GammaPath(Float64[], Float64[]))
+end
+
+struct TransitionDistribution{T}
+    current_state::MixedState{T}
+    model::LevyLangevin
+    s::T
+    t::T
+end
+
+function Base.rand(rng::Random.AbstractRNG, td::TransitionDistribution)
+    let model = td.model, s = td.s, t = td.t, state = td.current_state
+        dt = t - s
+        path = simulate(model.θ.process, dt, s, t, ϵ)
+        μ, Σ = meancov(t, model.θ.dyn, path, model.θ.nvm)
+        Σ += 1e-6 * I
+        return MixedState(
+            rand(rng, MultivariateNormal(exp(dyn, dt) * state.x + μ, Σ)), path
+        )
```

Comment:
Can't we just mutate the state instead ?

Reply:
That works in a purely filtering context, but if you then want to apply smoothing, you need all of the historic states saved, so mutating isn't an option.

I think this clunkiness is best solved by modifying AdvancedPS.jl to expect a sample to be returned rather than a distribution. I only wrote it this way to maintain compatibility.